### ISSUE TITLE
Unable to encrypt serialized array

### ISSUE THREAD

**Issue by 87cms** [phpseclib/phpseclib#601]
Hello,

I am implementing phpseclib on our project.
My test doesn't work with the following code : 

```php
$rijndael = new Crypt_Rijndael();
$rijndael->setKey( 'hellothisismykey' );
$array = array('k1' => 'value1', 'k2' => 'value2');
$serialized = serialize($array);
echo $rijndael->decrypt($rijndael->encrypt($serialized));
```

This code returns

```html
<b>Notice</b>:  Undefined offset: 3 in <b>phpseclib/Crypt/Base.php(1989) : runtime-created function</b> on line <b>319</b><br />

<b>Notice</b>:  Undefined offset: 4 in <b>phpseclib/Crypt/Base.php(1989) : runtime-created function</b> on line <b>320</b><br />
```

This code is working when I replace $serialized with a string.

Environment : Red Hat Linux 7 ; PHP 5.6.4 ; phpseclib/master from today

Thanks !

**Comment by terrafrost** [comment]
I'm not able to duplicate this but I'm not running 5.6.4 either. I'll try when I get home as time permits.

**Comment by terrafrost** [comment]
I can't reproduce this even on 5.6.4. php -v gives me this:

```
C:\some\directory>php56 -v
PHP 5.6.4 (cli) (built: Dec 17 2014 13:20:35)
Copyright (c) 1997-2014 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2014 Zend Technologies
```

I then try running this script and don't get any errors:

```
<?php
error_reporting(E_ALL);

include('Crypt/AES.php');

define('CRYPT_RIJNDAEL_MODE', CRYPT_MODE_INTERNAL);

$rijndael = new Crypt_Rijndael();
$rijndael->setKey( 'hellothisismykey' );
$array = array('k1' => 'value1', 'k2' => 'value2');
$serialized = serialize($array);
echo $rijndael->decrypt($rijndael->encrypt($serialized));
```

**Comment by terrafrost** [comment]
Here are the lines leading up to 319 / 320:

```php
                        $in = $_block = substr($_text, $_i, 16);
                        $in = unpack("N*", $in);
$s0 = $in[1] ^ 1528546082;
$s1 = $in[2] ^ -677214309;
$s2 = $in[3] ^ 1658105913;
$s3 = $in[4] ^ -841356084;
```

319 and 320 are the `$s2` and `$s3` lines.

**Comment by 87cms** [comment]
Thank you for your support !

I tried with your previous code, and I have a very strange behavior : 
- RHEL 7 : not working with Warning/Notices
- Debian 7 : working perfectly
Config and services installed are the same on this two servers.

So this "bug" maybe comes from OS and not from phpseclib.

**Comment by terrafrost** [comment]
`serialize(array('k1' => 'value1', 'k2' => 'value2'))` produces a 50-byte string. AES uses 16-byte blocks. 50 % 16 is 2. My guess is that if the ciphertext were a multiple of the block size it'd decrypt just fine. The question is...  why isn't it?

In Crypt/Base.php, there's this:

```php
$_text = str_pad($_text, strlen($_text) + ('.$block_size.' - strlen($_text) % '.$block_size.') % '.$block_size.', chr(0));
```
Replacing $block_size as appropriate yields this:


```php
$_text = str_pad($_text, strlen($_text) + (16 - strlen($_text) % 16) % 16, chr(0));
```

Subbing out strlen($_text):

```php
str_pad($_text, 50 + (16 - 2) % 16, chr(0));
```

The last % $block_size seems unnecessary but, basically, $_text is being null padded to 64 bytes and 64 is a multiple of 16.

So, basically, $_text should be 64 bytes long and that's if there wasn't padding. With padding it'd be 64-bytes as well, per this:

```php
$_text = $self->_pad($_text);
```

So in light of that can you tell me what the following outputs?:

```php
<?php
error_reporting(E_ALL);

include('Crypt/AES.php');

define('CRYPT_RIJNDAEL_MODE', CRYPT_MODE_INTERNAL);

$rijndael = new Crypt_Rijndael();
$rijndael->setKey( 'hellothisismykey' );
$array = array('k1' => 'value1', 'k2' => 'value2');
$serialized = serialize($array);
echo 'serialized = '.strlen($serialized)."\r\n";

$temp = $rijndael->encrypt($serialized);
echo 'temp = ' . strlen($temp) . "\r\n"'

echo $rijndael->decrypt($temp);
```

Thanks!

**Comment by terrafrost** [comment]
In Crypt/Base.php can you replace `_pad` with this?:

```php
    function _pad($text)
    {
echo "_pad called\r\n";
        $length = strlen($text);

        if (!$this->padding) {
            if ($length % $this->block_size == 0) {
echo "the text IS a multiple of the plaintext\r\n";
                return $text;
            } else {
echo "the text is NOT a multiple of the plaintext\r\n";
                user_error("The plaintext's length ($length) is not a multiple of the block size ({$this->block_size})");
                $this->padding = true;
            }
        }

        $pad = $this->block_size - ($length % $this->block_size);
echo "performing PKCS7 padding\r\n";
        return str_pad($text, $length + $pad, chr($pad));
    }
```

**Comment by 87cms** [comment]
Here is the result :

```html
serialized = 50
_pad called
performing PKCS7 padding
temp = 43
```

**Comment by terrafrost** [comment]
Thanks! Can you replace `_pad`, again, with this?:

```php
    function _pad($text)
    {
echo "_pad called\r\n";
        $length = strlen($text);

        if (!$this->padding) {
            if ($length % $this->block_size == 0) {
                return $text;
            } else {
                user_error("The plaintext's length ($length) is not a multiple of the block size ({$this->block_size})");
                $this->padding = true;
            }
        }

        $pad = $this->block_size - ($length % $this->block_size);
echo "performing PKCS7 padding\r\n";
echo "pad = $pad\r\n";
echo "block_size = {$this->block_size}\r\n";
echo "length = $length\r\n";
        $result = str_pad($text, $length + $pad, chr($pad));
echo "result = ".strlen($result)."\r\n";
return $result;
    }
```

Thanks!

**Comment by 87cms** [comment]
And here is the new output :

```html

serialized = 50
_pad called
performing PKCS7 padding
pad = 14
block_size = 16
length = 50
result = 64
temp = 43
<br />
<b>Notice</b>:  Undefined offset: 3 in <b>/var/www/lib/phpseclib/Crypt/Base.php(2011) : runtime-created function</b> on line <b>319</b><br />
<br />
<b>Notice</b>:  Undefined offset: 4 in <b>/var/www/lib/phpseclib/Crypt/Base.php(2011) : runtime-created function</b> on line <b>320</b><br />
```

**Comment by terrafrost** [comment]
In Crypt/Base.php can you find this?:

```php
            // case CRYPT_MODE_CBC:
            default:
                $encrypt = $init_encrypt . '
                    $_ciphertext = "";
                    $_text = $self->_pad($_text);
                    $_plaintext_len = strlen($_text);

                    $in = $self->encryptIV;

                    for ($_i = 0; $_i < $_plaintext_len; $_i+= '.$block_size.') {
                        $in = substr($_text, $_i, '.$block_size.') ^ $in;
                        '.$encrypt_block.'
                        $_ciphertext.= $in;
                    }

                    if ($self->continuousBuffer) {
                        $self->encryptIV = $in;
                    }

                    return $_ciphertext;
                    ';
```

...and replace it with this:

```php
            // case CRYPT_MODE_CBC:
            default:
                $encrypt = $init_encrypt . '
                    $_ciphertext = "";
                    $_text = $self->_pad($_text);
echo "_text_len = " . strlen($_text) . "\r\n";
                    $_plaintext_len = strlen($_text);

                    $in = $self->encryptIV;

                    for ($_i = 0; $_i < $_plaintext_len; $_i+= '.$block_size.') {
                        $in = substr($_text, $_i, '.$block_size.') ^ $in;
                        '.$encrypt_block.'
                        $_ciphertext.= $in;
                    }

                    if ($self->continuousBuffer) {
                        $self->encryptIV = $in;
                    }

echo "_ciphertext_len = ".strlen($_ciphertext)."\r\n";
                    return $_ciphertext;
                    ';
```

Thanks!

**Comment by 87cms** [comment]
```html

serialized = 50
_pad called
performing PKCS7 padding
pad = 14
block_size = 16
length = 50
result = 64
_text_len = 64
_ciphertext_len = 43
temp = 43
<br />
<b>Notice</b>:  Undefined offset: 3 in <b>/var/www/lib/phpseclib/Crypt/Base.php(2013) : runtime-created function</b> on line <b>321</b><br />
<br />
<b>Notice</b>:  Undefined offset: 4 in <b>/var/www/lib/phpseclib/Crypt/Base.php(2013) : runtime-created function</b> on line <b>322</b><br />
```

**Comment by terrafrost** [comment]
Try this:

```php
            // case CRYPT_MODE_CBC:
            default:
                $encrypt = $init_encrypt . '
                    $_ciphertext = "";
                    $_text = $self->_pad($_text);
echo "_text_len = " . strlen($_text) . "\r\n";
                    $_plaintext_len = strlen($_text);

                    $in = $self->encryptIV;

                    for ($_i = 0; $_i < $_plaintext_len; $_i+= '.$block_size.') {
                        $in = substr($_text, $_i, '.$block_size.') ^ $in;
                        '.$encrypt_block.'
echo "adding " . strlen($in) . " to ciphertext (" . strlen($_ciphertext) . ")\r\n";
                        $_ciphertext.= $in;
                    }

                    if ($self->continuousBuffer) {
                        $self->encryptIV = $in;
                    }

echo "_ciphertext_len = ".strlen($_ciphertext)."\r\n";
                    return $_ciphertext;
                    ';
```

**Comment by 87cms** [comment]
```html
serialized = 50
_pad called
performing PKCS7 padding
pad = 14
block_size = 16
length = 50
result = 64
_text_len = 64
adding 11 to ciphertext (0)
adding 9 to ciphertext (11)
adding 10 to ciphertext (19)
adding 14 to ciphertext (29)
_ciphertext_len = 43
temp = 43
<br />
<b>Notice</b>:  Undefined offset: 3 in <b>/var/www/lib/phpseclib/Crypt/Base.php(2014) : runtime-created function</b> on line <b>322</b><br />
<br />
<b>Notice</b>:  Undefined offset: 4 in <b>/var/www/lib/phpseclib/Crypt/Base.php(2014) : runtime-created function</b> on line <b>323</b><br />
```

**Comment by terrafrost** [comment]
In Crypt/Rijndael.php can you find this?:

```php
            $encrypt_block .= '$in = pack("N*"'."\n";
            for ($i = 0; $i < $Nb; ++$i) {
                $encrypt_block.= ',
                    ($'.$e.$i                  .' & 0xFF000000) ^
                    ($'.$e.(($i + $c[1]) % $Nb).' & 0x00FF0000) ^
                    ($'.$e.(($i + $c[2]) % $Nb).' & 0x0000FF00) ^
                    ($'.$e.(($i + $c[3]) % $Nb).' & 0x000000FF) ^
                    '.$w[$i]."\n";
            }
            $encrypt_block .= ');';
```

...and replace it with this?:

```php
            $encrypt_block .= '$in = pack("N*"'."\n";
$zzz = '';
            for ($i = 0; $i < $Nb; ++$i) {
                $temp = ',
                    ($'.$e.$i                  .' & 0xFF000000) ^
                    ($'.$e.(($i + $c[1]) % $Nb).' & 0x00FF0000) ^
                    ($'.$e.(($i + $c[2]) % $Nb).' & 0x0000FF00) ^
                    ($'.$e.(($i + $c[3]) % $Nb).' & 0x000000FF) ^
                    '.$w[$i]."\n";
$encrypt_block.=$temp;
$zzz.= 'echo "param[' . $i . ' / ' . $Nb . '] = ".('.trim($temp, ",\n") . ') . "\r\n";';
            }
            $encrypt_block .= ');';
$encrypt_block.= $zzz;
```

Thanks!

**Comment by 87cms** [comment]
Thanks for taking some time for this request.
I think it's working now  : 
```html

serialized = 50
_pad called
performing PKCS7 padding
pad = 14
block_size = 16
length = 50
result = 64
_text_len = 64
param[0 / 4] = 4092779460
param[1 / 4] = 925454021
param[2 / 4] = 3778223681
param[3 / 4] = 2329583302
adding 16 to ciphertext (0)
param[0 / 4] = 775874531
param[1 / 4] = 3745246956
param[2 / 4] = 2844776049
param[3 / 4] = 3939646365
adding 16 to ciphertext (16)
param[0 / 4] = 730785934
param[1 / 4] = 1059228756
param[2 / 4] = 4081141143
param[3 / 4] = 2429426877
adding 16 to ciphertext (32)
param[0 / 4] = 3871489125
param[1 / 4] = 4178030946
param[2 / 4] = 1629592885
param[3 / 4] = 808615072
adding 16 to ciphertext (48)
_ciphertext_len = 64
temp = 64
a:2:{s:2:"k1";s:6:"value1";s:2:"k2";s:6:"value2";}
```

**Comment by terrafrost** [comment]
I'm glad that it's working but the code changes I had you make weren't supposed to get it working. They echo'd out debug info but didn't change what the code is doing meh.

My working hypothesis had been that pack() on your system wasn't returning appropriately sized strings, which would indicate a PHP bug. But since it's now working, even tho I didn't actually change what the code was doing...  idk...

**Comment by 87cms** [comment]
I understand, I will ask if someone in your team has updated the VM template.

Here is the new result : 
```html

serialized = 50

                static $t0, $t1, $t2, $t3, $sbox;
                if (!$t0) {
                    for ($i = 0; $i < 256; ++$i) {
                        $t0[$i]    = (int)$self->t0[$i];
                        $t1[$i]    = (int)$self->t1[$i];
                        $t2[$i]    = (int)$self->t2[$i];
                        $t3[$i]    = (int)$self->t3[$i];
                        $sbox[$i]  = (int)$self->sbox[$i];
                    }
                }
            
                    $_ciphertext = "";
                    $_text = $self->_pad($_text);
                    $_plaintext_len = strlen($_text);

                    $in = $self->encryptIV;

                    for ($_i = 0; $_i < $_plaintext_len; $_i+= 16) {
                        $in = substr($_text, $_i, 16) ^ $in;
                        $in = unpack("N*", $in);
$s0 = $in[1] ^ 1751477356;
$s1 = $in[2] ^ 1869899881;
$s2 = $in[3] ^ 1936290669;
$s3 = $in[4] ^ 2037081465;
$e0 =
                        $t0[($s0 >> 24) & 0xff] ^
                        $t1[($s1 >> 16) & 0xff] ^
                        $t2[($s2 >>  8) & 0xff] ^
                        $t3[ $s3        & 0xff] ^
                        371776218;
$e1 =
                        $t0[($s1 >> 24) & 0xff] ^
                        $t1[($s2 >> 16) & 0xff] ^
                        $t2[($s3 >>  8) & 0xff] ^
                        $t3[ $s0        & 0xff] ^
                        2036118195;
$e2 =
                        $t0[($s2 >> 24) & 0xff] ^
                        $t1[($s3 >> 16) & 0xff] ^
                        $t2[($s0 >>  8) & 0xff] ^
                        $t3[ $s1        & 0xff] ^
                        171295198;
$e3 =
                        $t0[($s3 >> 24) & 0xff] ^
                        $t1[($s0 >> 16) & 0xff] ^
                        $t2[($s1 >>  8) & 0xff] ^
                        $t3[ $s2        & 0xff] ^
                        1935582375;
$s0 =
                        $t0[($e0 >> 24) & 0xff] ^
                        $t1[($e1 >> 16) & 0xff] ^
                        $t2[($e2 >>  8) & 0xff] ^
                        $t3[ $e3        & 0xff] ^
                        1281459797;
$s1 =
                        $t0[($e1 >> 24) & 0xff] ^
                        $t1[($e2 >> 16) & 0xff] ^
                        $t2[($e3 >>  8) & 0xff] ^
                        $t3[ $e0        & 0xff] ^
                        893203686;
$s2 =
                        $t0[($e2 >> 24) & 0xff] ^
                        $t1[($e3 >> 16) & 0xff] ^
                        $t2[($e0 >>  8) & 0xff] ^
                        $t3[ $e1        & 0xff] ^
                        1057551672;
$s3 =
                        $t0[($e3 >> 24) & 0xff] ^
                        $t1[($e0 >> 16) & 0xff] ^
                        $t2[($e1 >>  8) & 0xff] ^
                        $t3[ $e2        & 0xff] ^
                        1280725407;
$e0 =
                        $t0[($s0 >> 24) & 0xff] ^
                        $t1[($s1 >> 16) & 0xff] ^
                        $t2[($s2 >>  8) & 0xff] ^
                        $t3[ $s3        & 0xff] ^
                        4189085052;
$e1 =
                        $t0[($s1 >> 24) & 0xff] ^
                        $t1[($s2 >> 16) & 0xff] ^
                        $t2[($s3 >>  8) & 0xff] ^
                        $t3[ $s0        & 0xff] ^
                        3431819674;
$e2 =
                        $t0[($s2 >> 24) & 0xff] ^
                        $t1[($s3 >> 16) & 0xff] ^
                        $t2[($s0 >>  8) & 0xff] ^
                        $t3[ $s1        & 0xff] ^
                        4085619874;
$e3 =
                        $t0[($s3 >> 24) & 0xff] ^
                        $t1[($s0 >> 16) & 0xff] ^
                        $t2[($s1 >>  8) & 0xff] ^
                        $t3[ $s2        & 0xff] ^
                        3218328893;
$s0 =
                        $t0[($e0 >> 24) & 0xff] ^
                        $t1[($e1 >> 16) & 0xff] ^
                        $t2[($e2 >>  8) & 0xff] ^
                        $t3[ $e3        & 0xff] ^
                        2534242932;
$s1 =
                        $t0[($e1 >> 24) & 0xff] ^
                        $t1[($e2 >> 16) & 0xff] ^
                        $t2[($e3 >>  8) & 0xff] ^
                        $t3[ $e0        & 0xff] ^
                        1535120366;
$s2 =
                        $t0[($e2 >> 24) & 0xff] ^
                        $t1[($e3 >> 16) & 0xff] ^
                        $t2[($e0 >>  8) & 0xff] ^
                        $t3[ $e1        & 0xff] ^
                        2818936652;
$s3 =
                        $t0[($e3 >> 24) & 0xff] ^
                        $t1[($e0 >> 16) & 0xff] ^
                        $t2[($e1 >>  8) & 0xff] ^
                        $t3[ $e2        & 0xff] ^
                        399917681;
$e0 =
                        $t0[($s0 >> 24) & 0xff] ^
                        $t1[($s1 >> 16) & 0xff] ^
                        $t2[($s2 >>  8) & 0xff] ^
                        $t3[ $s3        & 0xff] ^
                        1898043780;
$e1 =
                        $t0[($s1 >> 24) & 0xff] ^
                        $t1[($s2 >> 16) & 0xff] ^
                        $t2[($s3 >>  8) & 0xff] ^
                        $t3[ $s0        & 0xff] ^
                        715246186;
$e2 =
                        $t0[($s2 >> 24) & 0xff] ^
                        $t1[($s3 >> 16) & 0xff] ^
                        $t2[($s0 >>  8) & 0xff] ^
                        $t3[ $s1        & 0xff] ^
                        2191803686;
$e3 =
                        $t0[($s3 >> 24) & 0xff] ^
                        $t1[($s0 >> 16) & 0xff] ^
                        $t2[($s1 >>  8) & 0xff] ^
                        $t3[ $s2        & 0xff] ^
                        2507278167;
$s0 =
                        $t0[($e0 >> 24) & 0xff] ^
                        $t1[($e1 >> 16) & 0xff] ^
                        $t2[($e2 >>  8) & 0xff] ^
                        $t3[ $e3        & 0xff] ^
                        300188334;
$s1 =
                        $t0[($e1 >> 24) & 0xff] ^
                        $t1[($e2 >> 16) & 0xff] ^
                        $t2[($e3 >>  8) & 0xff] ^
                        $t3[ $e0        & 0xff] ^
                        994396356;
$s2 =
                        $t0[($e2 >> 24) & 0xff] ^
                        $t1[($e3 >> 16) & 0xff] ^
                        $t2[($e0 >>  8) & 0xff] ^
                        $t3[ $e1        & 0xff] ^
                        3118534114;
$s3 =
                        $t0[($e3 >> 24) & 0xff] ^
                        $t1[($e0 >> 16) & 0xff] ^
                        $t2[($e1 >>  8) & 0xff] ^
                        $t3[ $e2        & 0xff] ^
                        747834037;
$e0 =
                        $t0[($s0 >> 24) & 0xff] ^
                        $t1[($s1 >> 16) & 0xff] ^
                        $t2[($s2 >>  8) & 0xff] ^
                        $t3[ $s3        & 0xff] ^
                        2374195167;
$e1 =
                        $t0[($s1 >> 24) & 0xff] ^
                        $t1[($s2 >> 16) & 0xff] ^
                        $t2[($s3 >>  8) & 0xff] ^
                        $t3[ $s0        & 0xff] ^
                        3066437403;
$e2 =
                        $t0[($s2 >> 24) & 0xff] ^
                        $t1[($s3 >> 16) & 0xff] ^
                        $t2[($s0 >>  8) & 0xff] ^
                        $t3[ $s1        & 0xff] ^
                        254219001;
$e3 =
                        $t0[($s3 >> 24) & 0xff] ^
                        $t1[($s0 >> 16) & 0xff] ^
                        $t2[($s1 >>  8) & 0xff] ^
                        $t3[ $s2        & 0xff] ^
                        599005260;
$s0 =
                        $t0[($e0 >> 24) & 0xff] ^
                        $t1[($e1 >> 16) & 0xff] ^
                        $t2[($e2 >>  8) & 0xff] ^
                        $t3[ $e3        & 0xff] ^
                        2150530809;
$s1 =
                        $t0[($e1 >> 24) & 0xff] ^
                        $t1[($e2 >> 16) & 0xff] ^
                        $t2[($e3 >>  8) & 0xff] ^
                        $t3[ $e0        & 0xff] ^
                        921199074;
$s2 =
                        $t0[($e2 >> 24) & 0xff] ^
                        $t1[($e3 >> 16) & 0xff] ^
                        $t2[($e0 >>  8) & 0xff] ^
                        $t3[ $e1        & 0xff] ^
                        969896731;
$s3 =
                        $t0[($e3 >> 24) & 0xff] ^
                        $t1[($e0 >> 16) & 0xff] ^
                        $t2[($e1 >>  8) & 0xff] ^
                        $t3[ $e2        & 0xff] ^
                        444296023;
$e0 =
                        $t0[($s0 >> 24) & 0xff] ^
                        $t1[($s1 >> 16) & 0xff] ^
                        $t2[($s2 >>  8) & 0xff] ^
                        $t3[ $s3        & 0xff] ^
                        3125880155;
$e1 =
                        $t0[($s1 >> 24) & 0xff] ^
                        $t1[($s2 >> 16) & 0xff] ^
                        $t2[($s3 >>  8) & 0xff] ^
                        $t3[ $s0        & 0xff] ^
                        2360951993;
$e2 =
                        $t0[($s2 >> 24) & 0xff] ^
                        $t1[($s3 >> 16) & 0xff] ^
                        $t2[($s0 >>  8) & 0xff] ^
                        $t3[ $s1        & 0xff] ^
                        3044423586;
$e3 =
                        $t0[($s3 >> 24) & 0xff] ^
                        $t1[($s0 >> 16) & 0xff] ^
                        $t2[($s1 >>  8) & 0xff] ^
                        $t3[ $s2        & 0xff] ^
                        2936888565;
$e0 =
                     $sbox[ $e0        & 0xff]        |
                    ($sbox[($e0 >>  8) & 0xff] <<  8) |
                    ($sbox[($e0 >> 16) & 0xff] << 16) |
                    ($sbox[($e0 >> 24) & 0xff] << 24);
$e1 =
                     $sbox[ $e1        & 0xff]        |
                    ($sbox[($e1 >>  8) & 0xff] <<  8) |
                    ($sbox[($e1 >> 16) & 0xff] << 16) |
                    ($sbox[($e1 >> 24) & 0xff] << 24);
$e2 =
                     $sbox[ $e2        & 0xff]        |
                    ($sbox[($e2 >>  8) & 0xff] <<  8) |
                    ($sbox[($e2 >> 16) & 0xff] << 16) |
                    ($sbox[($e2 >> 24) & 0xff] << 24);
$e3 =
                     $sbox[ $e3        & 0xff]        |
                    ($sbox[($e3 >>  8) & 0xff] <<  8) |
                    ($sbox[($e3 >> 16) & 0xff] << 16) |
                    ($sbox[($e3 >> 24) & 0xff] << 24);
$in = pack("N*"
,
                    ($e0 & 0xFF000000) ^
                    ($e1 & 0x00FF0000) ^
                    ($e2 & 0x0000FF00) ^
                    ($e3 & 0x000000FF) ^
                    1528546082
,
                    ($e1 & 0xFF000000) ^
                    ($e2 & 0x00FF0000) ^
                    ($e3 & 0x0000FF00) ^
                    ($e0 & 0x000000FF) ^
                    3617752987
,
                    ($e2 & 0xFF000000) ^
                    ($e3 & 0x00FF0000) ^
                    ($e0 & 0x0000FF00) ^
                    ($e1 & 0x000000FF) ^
                    1658105913
,
                    ($e3 & 0xFF000000) ^
                    ($e0 & 0x00FF0000) ^
                    ($e1 & 0x0000FF00) ^
                    ($e2 & 0x000000FF) ^
                    3453611212
);echo "param[0 / 4] = ".(                    ($e0 & 0xFF000000) ^
                    ($e1 & 0x00FF0000) ^
                    ($e2 & 0x0000FF00) ^
                    ($e3 & 0x000000FF) ^
                    1528546082) . "\r\n";echo "param[1 / 4] = ".(                    ($e1 & 0xFF000000) ^
                    ($e2 & 0x00FF0000) ^
                    ($e3 & 0x0000FF00) ^
                    ($e0 & 0x000000FF) ^
                    3617752987) . "\r\n";echo "param[2 / 4] = ".(                    ($e2 & 0xFF000000) ^
                    ($e3 & 0x00FF0000) ^
                    ($e0 & 0x0000FF00) ^
                    ($e1 & 0x000000FF) ^
                    1658105913) . "\r\n";echo "param[3 / 4] = ".(                    ($e3 & 0xFF000000) ^
                    ($e0 & 0x00FF0000) ^
                    ($e1 & 0x0000FF00) ^
                    ($e2 & 0x000000FF) ^
                    3453611212) . "\r\n";
                        $_ciphertext.= $in;
                    }

                    if ($self->continuousBuffer) {
                        $self->encryptIV = $in;
                    }

                    return $_ciphertext;
                    _pad called
performing PKCS7 padding
pad = 14
block_size = 16
length = 50
result = 64
param[0 / 4] = 4092779460
param[1 / 4] = 925454021
param[2 / 4] = 3778223681
param[3 / 4] = 2329583302
param[0 / 4] = 775874531
param[1 / 4] = 3745246956
param[2 / 4] = 2844776049
param[3 / 4] = 3939646365
param[0 / 4] = 730785934
param[1 / 4] = 1059228756
param[2 / 4] = 4081141143
param[3 / 4] = 2429426877
param[0 / 4] = 3871489125
param[1 / 4] = 4178030946
param[2 / 4] = 1629592885
param[3 / 4] = 808615072
temp = 64
a:2:{s:2:"k1";s:6:"value1";s:2:"k2";s:6:"value2";}
```html

**Comment by 87cms** [comment]
EDIT : FYI, a member of our team has installed imageMagick via pecl on the VM template. 
So I have tested with a fresh phpseclib 0.3.9, and all is working perfectly now !

Maybe your first intuition was right, it was a php bug.
Many thanks for your help !
